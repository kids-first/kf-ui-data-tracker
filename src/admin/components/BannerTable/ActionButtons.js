import React from 'react';
import {
  Icon,
  Divider,
  Responsive,
  Popup,
  Button,
  Table,
} from 'semantic-ui-react';

const ActionButtons = ({handleDelete, handleEdit}) => (
  <Table.Cell>
    <Button.Group basic size="small">
      {/* Edit Banner */}
      <Popup
        content="Edit banner"
        trigger={<Button icon="edit" onClick={(e) => handleEdit()} />}
      />
      {/* Delete Banner */}
      <Popup
        trigger={
          <Responsive
            as={Button}
            minWidth={Responsive.onlyTablet.minWidth}
            basic
            compact
            onClick={(e) => e.stopPropagation()}
            icon={<Icon name="trash alternate" />}
          />
        }
        header="Are you sure?"
        content={
          <>
            This banner will be deleted
            <Divider />
            <Button
              negative
              fluid
              icon={<Icon name="trash alternate" />}
              content="Delete"
              onClick={(e) => {
                e.stopPropagation();
                handleDelete();
              }}
            />
          </>
        }
        on="click"
        position="top right"
      />
      ;
    </Button.Group>
  </Table.Cell>
);

export default ActionButtons;
