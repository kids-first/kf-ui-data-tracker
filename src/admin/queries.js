import gql from 'graphql-tag';
import {TOKEN_FIELDS} from './fragments';
import {
  STUDY_BASIC_FIELDS,
  CREATOR_FIELDS,
  USER_FIELDS,
  GROUP_FIELDS,
} from '../state/fragments';

// Query to get developer tokens
export const GET_DEV_TOKENS = gql`
  query DevTokens {
    allDevTokens {
      edges {
        node {
          ...TokenFields
        }
      }
    }
  }
  ${TOKEN_FIELDS}
`;

// Get all buckets
export const GET_BUCKETS = gql`
  query Buckets {
    allBuckets {
      edges {
        node {
          id
          name
          deleted
          createdOn
          study {
            ...StudyBasicFields
          }
        }
      }
    }
  }
  ${STUDY_BASIC_FIELDS}
`;

// All pending invites
export const ALL_REFERRAL_TOEKNS = gql`
  query allReferralTokens {
    allReferralTokens {
      edges {
        node {
          id
          email
          claimed
          isValid
          createdAt
          createdBy {
            ...CreatorFields
          }
          claimedBy {
            dateJoined
            ...CreatorFields
          }
          studies {
            edges {
              node {
                ...StudyBasicFields
              }
            }
          }
          groups {
            edges {
              node {
                id
                name
              }
            }
          }
        }
      }
    }
  }
  ${CREATOR_FIELDS}
  ${STUDY_BASIC_FIELDS}
`;

// Get all users with their groups
export const ALL_USERS = gql`
  query AllUsers {
    allUsers {
      edges {
        node {
          id
          ...UserFields
          ...GroupFields
        }
      }
    }
  }
  ${USER_FIELDS}
  ${GROUP_FIELDS}
`;

export const ALL_LOGS = gql`
  query AllLogs {
    allJobLogs(orderBy: "-created_at", first: 10) {
      edges {
        node {
          id
          createdAt
          downloadUrl
          job {
            id
            name
          }
        }
      }
    }
  }
`;

export const GET_LOG = gql`
  query Log($id: ID!) {
    jobLog(id: $id) {
      id
      createdAt
      downloadUrl
      job {
        id
        name
      }
    }
  }
`;
