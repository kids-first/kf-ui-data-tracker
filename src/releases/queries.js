import gql from 'graphql-tag';
import {RELEASE_FIELDS, SERVICE_FIELDS} from './fragments';

export const GET_RELEASES = gql`
  query AllReleases {
    allReleases(first: 20, orderBy: "-created_at") {
      edges {
        node {
          ...ReleaseFields
          tasks {
            edges {
              node {
                releaseService {
                  ...ServiceFields
                }
              }
            }
          }
        }
      }
    }
  }
  ${RELEASE_FIELDS}
  ${SERVICE_FIELDS}
`;

export const LATEST_RELEASE = gql`
  query LatestRelease {
    allReleases(first: 1, orderBy: "-created_at", state: "published") {
      edges {
        node {
          ...ReleaseFields
        }
      }
    }
  }
  ${RELEASE_FIELDS}
`;

export const ALL_SERVICES = gql`
  query AllServices {
    allReleaseServices(first: 20, orderBy: "-created_at") {
      edges {
        node {
          ...ServiceFields
        }
      }
    }
  }
  ${SERVICE_FIELDS}
`;

export const GET_SERVICE = gql`
  query GetService($id: ID!) {
    releaseService(id: $id) {
      ...ServiceFields
    }
  }
  ${SERVICE_FIELDS}
`;

export const ALL_EVENTS = gql`
  query AllEvents($release: ID, $first: Int, $releaseService: ID) {
    allReleaseEvents(
      release: $release
      releaseService: $releaseService
      orderBy: "-created_at"
      first: $first
    ) {
      edges {
        node {
          id
          createdAt
          message
          release {
            id
            kfId
            name
          }
          releaseService {
            id
            kfId
            name
          }
          task {
            id
            kfId
          }
        }
      }
    }
  }
`;

export const ALL_TASKS = gql`
  query AllTasks($first: Int, $release: ID!) {
    allReleaseTasks(first: $first, release: $release) {
      edges {
        node {
          id
          kfId
          state
          createdAt
          releaseService {
            id
            name
          }
        }
      }
    }
  }
`;

export const ALL_NOTES = gql`
  query AllNotes($release: ID) {
    allReleaseNotes(release: $release) {
      edges {
        node {
          id
          kfId
          description
          createdAt
          study {
            id
          }
        }
      }
    }
  }
`;

export const GET_RELEASE = gql`
  query GetRelease($id: ID!) {
    release(id: $id) {
      ...ReleaseFields
      studies {
        edges {
          node {
            id
            kfId
            name
            createdAt
          }
        }
      }
      tasks {
        edges {
          node {
            id
            kfId
            state
            jobLog {
              id
              downloadUrl
            }
            releaseService {
              ...ServiceFields
            }
          }
        }
      }
    }
  }
  ${RELEASE_FIELDS}
  ${SERVICE_FIELDS}
`;
