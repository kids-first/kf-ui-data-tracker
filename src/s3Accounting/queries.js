import {FILE_AUDI_FIELDS} from './fragments';
import gql from 'graphql-tag';

export const ALL_STORAGE_ANALYSES = gql`
  query allStorageAnalyses($studyKfId: String) {
    allStorageAnalyses(studyKfId: $studyKfId) {
      edges {
        node {
          id
          uuid
          createdAt
          refreshedAt
          scannedStorageAt
          stats
          study {
            id
            kfId
            name
          }
          fileAudits {
            edges {
              node {
                ...FileAudiFields
              }
            }
          }
        }
      }
    }
  }
  ${FILE_AUDI_FIELDS}
`;

export const ALL_FILE_AUDITS = gql`
  query allFileAudits(
    $storageAnalysis: ID
    $result: ResultEnum
    $first: Int
    $last: Int
    $after: String
    $before: String
  ) {
    allFileAudits(
      storageAnalysis: $storageAnalysis
      result: $result
      first: $first
      last: $last
      after: $after
      before: $before
      orderBy: "-created_at"
    ) {
      pageInfo {
        hasNextPage
        hasPreviousPage
        startCursor
        endCursor
      }
      edges {
        node {
          ...FileAudiFields
        }
      }
    }
  }
  ${FILE_AUDI_FIELDS}
`;

export const ALL_CLOUD_INVENTORY_FILES = gql`
  query allCloudInventoryFiles($storageAnalysis: ID, $result: ResultEnum) {
    allCloudInventoryFiles(storageAnalysis: $storageAnalysis, result: $result) {
      edges {
        node {
          ...FileAudiFields
        }
      }
      pageInfo {
        hasNextPage
        hasPreviousPage
        startCursor
        endCursor
      }
    }
  }
  ${FILE_AUDI_FIELDS}
`;

export const ALL_UPLOAD_MANIFEST_FILES = gql`
  query allUploadManifestFiles($storageAnalysis: ID, $result: ResultEnum) {
    allUploadManifestFiles(storageAnalysis: $storageAnalysis, result: $result) {
      edges {
        node {
          ...FileAudiFields
        }
      }
      pageInfo {
        hasNextPage
        hasPreviousPage
        startCursor
        endCursor
      }
    }
  }
  ${FILE_AUDI_FIELDS}
`;
