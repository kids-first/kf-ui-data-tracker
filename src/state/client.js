import {ApolloClient, HttpLink} from '@apollo/client';
import {createUploadLink} from 'apollo-upload-client';
import {InMemoryCache} from '@apollo/client/cache';
import {KF_STUDY_API, KF_COORD_API} from '../common/globals';

const link = new HttpLink.split(
  operation => operation.getContext().clientName === 'coordinator',
  new HttpLink({uri: `${KF_COORD_API}/graphql`}),
  new createUploadLink({uri: `${KF_STUDY_API}/graphql`}),
);

const policies = {
  typePolicies: {
    StudyNode: {
      fields: {
        kfId: {
          merge(existing, incoming) {
            console.log('merging');
            // Better, but not quite correct.
            return {...existing, ...incoming};
          },
        },
      },
    },
  },
};

export const client = new ApolloClient({
  cache: new InMemoryCache(),
  headers: {
    authorization: localStorage.getItem('accessToken')
      ? `Bearer ${localStorage.getItem('accessToken')}`
      : '',
  },
  link,
});
