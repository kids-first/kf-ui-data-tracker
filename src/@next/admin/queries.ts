import {gql} from '@apollo/client';

import {USER_FIELDS} from './fragments';

export const ALL_USERS = gql`
    query AllUsers($before: String, $after: String, $first: Int, $last: Int) {
        allUsers(before: $before, after: $after, first: $first, last: $last) {
            pageInfo {
                startCursor
                endCursor
                hasNextPage
                hasPreviousPage
            }
            edges {
                cursor
                node {
                    id
                    ...UserFields
                }
            }
        }
    }
    ${USER_FIELDS}
`;
