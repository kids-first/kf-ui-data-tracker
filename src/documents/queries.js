import gql from 'graphql-tag';
import {FILE_FIELDS, VERSION_FIELDS} from './fragments';
import {CREATOR_FIELDS} from '../state/fragments';

// Query to get a file by its kf id
export const GET_FILE_BY_ID = gql`
  query File($kfId: String!) {
    fileByKfId(kfId: $kfId) {
      ...FileFields
      creator {
        ...CreatorFields
      }
      templateVersion {
        id
        createdAt
        modifiedAt
        dataTemplate {
          id
          name
          description
          icon
        }
      }
      versions {
        edges {
          node {
            analysis {
              id
              columns
              nrows
              ncols
              knownFormat
              errorMessage
            }
            ...VersionFields
            downloadUrl
            creator {
              ...CreatorFields
            }
          }
        }
      }
    }
  }
  ${FILE_FIELDS}
  ${VERSION_FIELDS}
  ${CREATOR_FIELDS}
`;
