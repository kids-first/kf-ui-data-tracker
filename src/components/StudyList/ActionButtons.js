import React, {useState} from 'react';
import {Link} from 'react-router-dom';
import {useQuery, useMutation} from '@apollo/react-hooks';
import {ADD_COLLABORATOR, CREATE_REFERRAL_TOKEN} from '../../state/mutations';
import {GET_STUDY_BY_ID, MY_PROFILE} from '../../state/queries';
import {Amplitude} from '@amplitude/react-amplitude';
import {Button, Label, Icon, Popup, Table} from 'semantic-ui-react';
import {
  countStudyNotification,
  countProjectNotification,
} from '../../common/notificationUtils';
import {hasPermission} from '../../common/permissions';
import CavaticaLogo from '../../assets/CavaticaLogo';
import AddCollaboratorModal from '../../modals/AddCollaboratorModal';

/**
 * A button that displays a popup when hovered and optionally a notification
 * label in the right corner.
 */
const PopupButton = ({
  header,
  content,
  icon,
  label = null,
  hoverable = false,
  disabled = false,
  ...props
}) => (
  <Amplitude
    eventProperties={inheritedProps => ({
      ...inheritedProps,
      scope: inheritedProps.scope
        ? [...inheritedProps.scope, 'button', header.toLowerCase()]
        : ['button', header.toLowerCase()],
    })}
  >
    {({logEvent}) => (
      <Popup
        disabled={disabled}
        header={header}
        content={content}
        hoverable={hoverable}
        position="top right"
        trigger={
          <Button
            icon={icon}
            style={{position: 'relative'}}
            content={label}
            onClick={() => logEvent('click')}
            {...props}
          />
        }
      />
    )}
  </Amplitude>
);
/**
 * Contains a button group for study actions
 */
const ActionButtons = ({study}) => {
  const [showCollaborators, setShowCollaborators] = useState(false);

  // Get current user's profile to determine permission level
  const {data: profileData, loading} = useQuery(MY_PROFILE);
  const user = profileData && profileData.myProfile;

  const studyInfoNotif = countStudyNotification(study);
  const projectNotif = countProjectNotification(study);

  const [addCollaborator] = useMutation(ADD_COLLABORATOR, {
    refetchQueries: [
      {
        query: GET_STUDY_BY_ID,
        variables: {
          kfId: study.kfId,
        },
      },
    ],
  });
  const [createToken] = useMutation(CREATE_REFERRAL_TOKEN);

  if (loading) return <span>loading</span>;

  return (
    <Amplitude
      eventProperties={inheritedProps => ({
        ...inheritedProps,
        scope: inheritedProps.scope
          ? [...inheritedProps.scope, 'study actions']
          : ['study actions'],
      })}
    >
      <Table.Cell singleLine width="1">
        <Button.Group icon basic>
          <PopupButton
            header="Study Information"
            icon="info"
            as={Link}
            to={`/study/${study.kfId}/basic-info/info`}
            content={
              hasPermission(user, 'change_study') &&
              studyInfoNotif > 0 && (
                <p>
                  <Icon name="warning sign" /> Missing {studyInfoNotif} fields
                </p>
              )
            }
            label={
              hasPermission(user, 'change_study') &&
              studyInfoNotif > 0 && (
                <Label empty corner circular color="orange" />
              )
            }
          />
          <PopupButton
            header="Documents"
            icon="file"
            as={Link}
            to={`/study/${study.kfId}/documents`}
          />
          <PopupButton
            header="Cavatica Projects"
            icon={<CavaticaLogo className="mr-5 vertical-middle" />}
            as={Link}
            to={`/study/${study.kfId}/cavatica`}
            content={
              hasPermission(user, 'link_project') &&
              projectNotif > 0 && (
                <p>
                  <Icon name="warning sign" /> Missing {projectNotif} required
                  projects
                </p>
              )
            }
            label={
              hasPermission(user, 'link_project') &&
              projectNotif > 0 && <Label empty corner circular color="orange" />
            }
          />
          <PopupButton
            hoverable
            header="Collaborators"
            icon="users"
            content={
              hasPermission(user, 'add_collaborator') && (
                <>
                  <p>Manage collaborators in this study</p>
                  <Button
                    primary
                    fluid
                    content="Add Collaborator"
                    icon="add"
                    labelPosition="left"
                    onClick={() => setShowCollaborators(true)}
                  />
                </>
              )
            }
            as={Link}
            to={`/study/${study.kfId}/collaborators`}
          />
        </Button.Group>
      </Table.Cell>
      <AddCollaboratorModal
        study={study}
        open={showCollaborators}
        onCloseDialog={() => setShowCollaborators(false)}
        addCollaborator={addCollaborator}
        inviteCollaborator={createToken}
      />
    </Amplitude>
  );
};

export default ActionButtons;
