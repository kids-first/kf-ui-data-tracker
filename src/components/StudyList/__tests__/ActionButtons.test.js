import React from 'react';
import {MemoryRouter} from 'react-router-dom';
import {MockedProvider} from '@apollo/client/testing';
import {render} from '@testing-library/react';
import wait from 'waait';
import {MY_PROFILE} from '../../../../src/state/queries';
import myProfile from '../../../../__mocks__/kf-api-study-creator/responses/myProfile.json';
import ActionButtons from '../ActionButtons';
import allStudies from '../../../../__mocks__/kf-api-study-creator/responses/allStudies.json';

it('release version correctly', async () => {
  const myProfileMock = {
    request: {
      query: MY_PROFILE,
    },
    result: myProfile,
  };

  const tree = render(
    <MockedProvider mocks={[myProfileMock]}>
      <MemoryRouter>
        <ActionButtons study={allStudies.data.allStudies.edges[0].node} />
      </MemoryRouter>
    </MockedProvider>,
  );

  // Loading state
  expect(tree.container).toMatchSnapshot();
  await wait(0);
  // Normal state
  expect(tree.container).toMatchSnapshot();
});
