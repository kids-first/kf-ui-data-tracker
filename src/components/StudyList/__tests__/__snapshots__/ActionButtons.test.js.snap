// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`release version correctly 1`] = `
<div>
  <td
    class="single line one wide"
  >
    <button
      class="ui basic loading disabled button"
      disabled=""
      tabindex="-1"
    >
      Loading
    </button>
  </td>
</div>
`;

exports[`release version correctly 2`] = `
<div>
  <td
    class="single line one wide"
  >
    <button
      class="ui basic loading disabled button"
      disabled=""
      tabindex="-1"
    >
      Loading
    </button>
  </td>
</div>
`;
